[기능 추가] 결과 기능

chech2 opened the pull request:
## 작업 내용
결과기능 CRU

## 테스트
### Repository
<img width="738" alt="스크린샷 2024-08-27 오후 2 25 36" src="https://github.com/user-attachments/assets/92855e35-3414-49c1-8b91-aa6042e6a562">
<img width="759" alt="스크린샷 2024-08-27 오전 8 45 59" src="https://github.com/user-attachments/assets/0fb9127b-fe0c-4c0b-9dd8-77dd97d9a11a">



### Service
<img width="756" alt="스크린샷 2024-08-27 오후 1 46 04" src="https://github.com/user-attachments/assets/95f0b472-60d4-43f1-8cfc-34c2b7e6432b">
<img width="761" alt="스크린샷 2024-08-27 오후 1 45 51" src="https://github.com/user-attachments/assets/278faa9d-3c29-4675-b746-7b7aa8023fb2">


### Controller
- 멤버 record 생성
<img width="1437" alt="스크린샷 2024-08-27 오후 9 55 20" src="https://github.com/user-attachments/assets/f4a0eb9f-bf55-46ae-8ee9-5eaf5a8b55a6">

- 멤버 결과 조회
<img width="1429" alt="스크린샷 2024-08-27 오후 9 59 10" src="https://github.com/user-attachments/assets/730b771a-a126-48b5-ad82-ecb0e402af51">
<img width="1449" alt="스크린샷 2024-08-27 오후 9 59 27" src="https://github.com/user-attachments/assets/249d6478-7914-44a1-a82a-e617614b3c7f">


- 게스트 record 생성
<img width="1443" alt="스크린샷 2024-08-27 오후 10 20 34" src="https://github.com/user-attachments/assets/1b465acd-0904-410a-aeb8-cda3c8571e6a">

- 게스트 결과 조회
<img width="1411" alt="스크린샷 2024-08-27 오후 10 21 14" src="https://github.com/user-attachments/assets/41ca0deb-880d-488c-9b09-88ae40c086ad">
<img width="1385" alt="gsRecordGET" src="https://github.com/user-attachments/assets/c799087e-470a-48de-8152-0feb06f7e9d9">

## 기타 사항 (참고 자료, 문의 사항 등)
- 예외처리는 새로운 브랜치에서 작업하겠습니다.
- 빠진 부분 있으면 말씀해주세요.
- 결과 상태 관련 기능은 다른 브랜치에서 작업하겠습니다.

chech2 committed:
[REFACTOR] Guest클래스에 NicknameProvider 인터페이스 적용 (#75)
- NicknameList에서 보다 명확한 의미 구분을 위해 추가

chech2 committed:
[FEAT] GSRecordController 추가 (#75)

chech2 committed:
[FEAT] GSRecordService 추가 (#75)

chech2 committed:
[REFACTOR] GuestScheduleRecord 시간 선택 컬럼 변경 (#75)
- start_date, end_date를 selectTime로 변경

chech2 committed:
[FEAT] TmpGSRepository, TmpGuestRepository, GSRecordRepository 추가 (#75)

chech2 committed:
[FEAT] GSRecordGetResponse 추가 (#75)

chech2 committed:
[FEAT] GSRecordPostRequest 추가 (#75)

chech2 committed:
[FEAT] NicknameProvider, NicknameList 추가 (#75)

chech2 committed:
[FEAT] SelectionDetail 추가 (#75)

chech2 committed:
[FEAT] ParticipantDetails 추가 (#75)

chech2 committed:
[FEAT] GSRecordSelectionDetail 추가 (#75)

chech2 committed:
[FEAT] GSRecordNicknameList 추가 (#75)

chech2 committed:
[FEAT] GSRecordException 추가 (#75)

chech2 committed:
[FEAT] GSRecordErrorCode 추가 (#75)

chech2 committed:
[REFACTOR] 예외처리 방식 변경, 메소드 명 변경 (#75)

chech2 committed:
[REFACTOR] 변경된 메소드 명 적용 (#75)

chech2 committed:
[REFACTOR] 선택시간에 대한 검증 추가 (#75)

chech2 committed:
[REFACTOR] 기존 record정보 삭제 (#75)
- post 요청 보내는 guest의 기존 record 정보 삭제 메소드 추가

chech2 committed:
[REFACTOR] 기존 record정보 삭제 (#75)
- guestId를 통한 기존 record 정보 삭제 메소드 추가
- response 변수명 변경

chech2 committed:
[REFACTOR] 누락된 final, 컨벤션에 맞춰 수정 (#75)

chech2 committed:
[REFACTOR] 누락된 final, 컨벤션에 맞춰 수정 (#75)
- controller

chech2 committed:
[REFACTOR] 예외처리 메세지 수정 (#75)

chech2 committed:
[REFACTOR] MemberScheduleRecord 시간관련 멤버 변수 변경 (#75)
- startDate, endDate -> selectTime으로 변경

chech2 committed:
[REFACTOR] GSRecordController post요청 리턴 변경 (#75)

chech2 committed:
[REFACTOR] 누락된 정적 팩토리 메소드 적용 (#75)
- 생성자 접근 제한자 변경: public -> private

chech2 committed:
[REFACTOR] GSRecordService 리팩토링 (#75)
- Transactional 어노테이션 사용
- 누락된 정적 팩토리 메소드 사용
- 예외 메세지 수정

chech2 committed:
[REFACTOR] Member 클래스에 NicknameProvider 인터페이스 적용 (#75)
- NicknameList에서 보다 명확한 의미 구분을 위해 추가

chech2 committed:
[FEAT] MSRecordController 추가 (#75)

chech2 committed:
[FEAT] MSRecordService 추가 (#75)

chech2 committed:
[FEAT] MSRecordSelectionDetail 추가 (#75)

chech2 committed:
[FEAT] MSRecordPostRequest 추가 (#75)

chech2 committed:
[FEAT] MSRecordNicknameList 추가 (#75)

chech2 committed:
[FEAT] MSRecordGetRequest, MSRecordGetResponse 추가 (#75)

chech2 committed:
[FEAT] TmpRoomMemberRepository, TmpMSRepository, TmpMemberRepository 추가 (#75)
- develop 브랜치 머지 전 기능 사용을 위해 임시로 사용한 tmprepository들 추가

chech2 committed:
[FEAT] MSRecordRepository 추가 (#75)

chech2 committed:
[REFACTOR] findByGuestScheduleId메소드 param 수정 (#75)

chech2 committed:
Merge remote-tracking branch 'origin/develop' into FEAT/HM-75

# Conflicts:
#	src/main/java/org/chzzk/howmeet/domain/common/embedded/date/impl/ScheduleDate.java
#	src/main/java/org/chzzk/howmeet/domain/regular/auth/entity/Member.java

chech2 committed:
[REFACTOR] 임시로 사용한 repository 제거 (#75)
- develop 브랜치와 머지 전 임시로 사용한 repository 제거: TmpRoomMemberRepository, TmpMSRepository, TmpMemberRepository, TmpGSRepository

chech2 committed:
[REFACTOR] TmpGSRepository 사용 로직 변경, GuestSchedule dates, time사용 (#75)
- TmpGSRepository -> GSRepository
- LocalDateTime으로 사용한 startDate, endDate대신 startTime, endTime 사용

chech2 committed:
[REFACTOR] MSRecordService 로직 수  (#75)
- TmpMSRepository,TmpMemberRepository, TmpRoomMemberRepository 삭제
- LocalDateTime으로 사용한 startDate, endDate 대신 startTime, endTime
- findMemberByRoomId메소드 추가

chech2 committed:
[REFACTOR] NicknameProvider 인터페이스 추가 (#75)

chech2 committed:
Merge remote-tracking branch 'origin/develop' into FEAT/HM-75

# Conflicts:
#	src/main/java/org/chzzk/howmeet/domain/regular/record/entity/MemberScheduleRecord.java

chech2 committed:
[REFACTOR] @Getter 삭제 (#75)
- 중복되어 작성한 부분이어서 삭제

chech2 committed:
[REFACTOR] import문 오류 수정 (#75)

chech2 committed:
[FEAT] GSRecord 테스트 코드 추가 (#75)
- GSRecordRepositoryTest
- GSRecordServiceTest
- TestConfig
- GSRecordFixture

chech2 committed:
[REFACTOR] RoomId 컬럼 추가 (#75)

chech2 committed:
[REFACTOR] RoomName 추가 (#75)

chech2 committed:
[REFACTOR] RoomRepository 추가 (#75)
- RoomRepository 추가
- findRoomByRoomId 메소드 추가

chech2 committed:
[FEAT] MSRecord 테스트 코드 추가 (#75)
- TestConfig
- MSRecordRepositoryTest
- MSRecordServiceTest

chech2 committed:
Merge remote-tracking branch 'origin/develop' into FEAT/HM-75

chech2 committed:
[REFACTOR] roomID 컬럼 명 수정 (#75)
- 생성 시 roomId 추가를 위한 생성자, of 메소드 수정

kmw2378 commented on src/main/java/org/chzzk/howmeet/domain/common/model/NicknameList.java:
```diff
@@ -0,0 +1,46 @@
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
+public class NicknameList {
```
클래스명에 자료형을 명시하기보단 `Nicknames` 와 같은 복수형은 어떨까요?

chech2 replied:
클래스명 고민하다가 생각이 안나서 NicknameList로 작성했었는데 복수형이 더 좋은 것 같네요! 수정했습니다!

kmw2378 commented on src/main/java/org/chzzk/howmeet/domain/common/model/NicknameList.java:
```diff
@@ -0,0 +1,46 @@
+
+    @JsonValue
+    public List<String> getNicknames() {
+        return nicknames;
```
불변성이 지켜지지 않을 것 같아요! `List.copyOf()`를 쓰는건 어떨까요?
추가로, 굳이 Getter가 필요하지 않다면 제공하지 않는게 좋을 것 같습니다!

chech2 replied:
굳이 필요하지 않을 것 같아 제공하지 않는 방향으로 수정했습니다

kmw2378 commented on src/main/java/org/chzzk/howmeet/domain/common/model/SelectionDetail.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    public static List<SelectionDetail> convertMapToSelectionDetailsList(
+            final HashMap<LocalDateTime, ? extends NicknameList> selectTimeMap) {
```
파라미터 타입으로 HashMap 대신 `Map` 인터페이스를 쓰는건 어떨까요? 추후, 동시성 이슈로 인해 `ConcurrentHashMap<>` 등으로 바뀔 수 있을 것 같아서요!

chech2 replied:
좋은 의견 감사합니다! 수정했어요!

kmw2378 commented on src/main/java/org/chzzk/howmeet/domain/regular/record/controller/MSRecordController.java:
```diff
@@ -0,0 +1,37 @@
+@RestController
+public class MSRecordController {
+
+    final MSRecordService msRecordService;
```
`private` 접근 지정자가 누락되었습니다!

chech2 replied:
수정했습니다! 감사합니다!

chech2 committed:
[REFACTOR] postGSRecord, postMSRecord 메소드 응답 변경 (#75)

kmw2378 commented on src/main/java/org/chzzk/howmeet/domain/regular/record/model/MSRecordNicknameList.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    public static NicknameList convertMapToNickNameList(final List<MemberScheduleRecord> msRecords,
+            final Map<Long, Nickname> nickNameMap) {
```
`nicknameMap` 이름이 직관적이지 않은 것 같아요! PK를 Key로 하는 Map인 것 같으니 `nicknamesById`와 같은 이름은 어떨까요?

chech2 replied:
수정했습니다.

kmw2378 commented on src/main/java/org/chzzk/howmeet/domain/regular/record/model/MSRecordNicknameList.java:
```diff
@@ -0,0 +1,32 @@
+    public static NicknameList convertMapToNickNameList(final List<MemberScheduleRecord> msRecords,
+            final Map<Long, Nickname> nickNameMap) {
+        NicknameList nicknameList = create();
+        HashSet<Nickname> nickNameSet = new HashSet<>();
```
- 일정 참가자간 중복 닉네임이 있어서 `Set<>` 을 사용하시는걸까요? 그렇다면 중복을 허용해서 제공해주는게 UX적으로 더 좋다고 생각합니다!
- 타입을 `Set<>` 으로 해도 좋을 것 같아요!
- 변수명에 자료형을 포함시키는 대신 `distinctNicknames` 와 같은 작명은 어떨까요?

chech2 replied:
일정을 선택한 회원 닉네임을 구하는 로직으로 1명의 회원이 여러 개의 선택했을 때 닉네임 중복을 없애기 위해 set을 사용했습니다.
그런데 회원의 경우, 중복 닉네임에 대해 허용을 해줘야 하기 때문에 로직 변경이 필요할 것 같네요. 
이 부분은 리팩토링 때 수정해서 pr 올리도록 하겠습니다. 

타입과 메소드 명은 수정했습니다. 메소드명 정할 때 고민하다가 작성했는데 좋은 의견 감사합니다!

kmw2378 commented on src/main/java/org/chzzk/howmeet/domain/regular/record/model/MSRecordNicknameList.java:
```diff
@@ -0,0 +1,32 @@
+        }
+        nicknameList.addFromNicknameSet(nickNameSet);
+
+        return nicknameList;
```
`stream()` 을 사용하면 더 간단해질 것 같아요!

```java
final List<Nickname> nicknames = msRecords.stream()
	.map(MemberScheduleRecord::getMemberId)
	.map(nicknameMap::get)
	.distinct()
	.toList();

return new NicknameList(nicknames);
```

chech2 replied:
수정완료했습니다

kmw2378 commented on src/main/java/org/chzzk/howmeet/domain/regular/record/model/MSRecordSelectionDetail.java:
```diff
@@ -0,0 +1,33 @@
+            selectTimeMap.computeIfAbsent(selectTime, k -> MSRecordNicknameList.create()).add(nickname);
+        }
+        return convertMapToSelectionDetailsList(selectTimeMap);
+    }
```
공통 로직이라면 `SelectionDetails`에 둬도 좋을 것 같아요! 

만약, `msRecords` 라는 인수가 문제라면 `MemberScheduleRecord` 와 `GuestScheduleRecord` 에 공통 인터페이스를 두면 될 것 같습니다!

chech2 replied:
넵 이 부분은 저번에 말씀 드렸던 부분이랑 같이 리팩토링 예정입니다!!
좋은 의견 주셔서 감사드립니다!

kmw2378 commented on src/main/java/org/chzzk/howmeet/domain/regular/record/service/MSRecordService.java:
```diff
@@ -0,0 +1,153 @@
+    final MSRecordRepository msRecordRepository;
+    final MSRepository msRepository;
+    final RoomMemberRepository roomMemberRepository;
+    final RoomRepository roomRepository;
```
`private`이 누락되있습니다!

chech2 replied:
수정했습니다! 감사합니다!

kmw2378 commented on src/main/java/org/chzzk/howmeet/domain/temporary/record/repository/TmpGuestRepository.java:
```diff
@@ -0,0 +1,11 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.repository.query.Param;
+
+public interface TmpGuestRepository extends JpaRepository<Guest, Long> {
+    List<Guest> findByGuestScheduleId(@Param("guestScheduleId") final Long guestScheduleId);
+
+}
```
`GuestRepository`를 사용하지 않은 이유가 궁금합니다!

chech2 replied:
해당 메소드가 GuestRepository에 존재하지 않았습니다.
그래서 해당 메소드는 추가 요청 드렸고 임시로 TmpGuestRepository 만들어서 사용했습니다.

kmw2378 reviewed (commented):
수고하셨습니다! 내일 마저 리뷰 남기겠습니다ㅜㅜ

- 일급 컬렉션 클래스명에 자료형을 넣는 대신 복수형을 사용하는 건 어떨까요?
- 변수 앞에 `final` 키워드, 필드 앞에 `private` 이 생략된 코드가 많습니다! 수정 부탁드립니다!

SSUHYUNKIM commented on src/main/java/org/chzzk/howmeet/domain/regular/record/controller/MSRecordController.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    @GetMapping
+    public ResponseEntity<?> getMSRecord(@RequestBody final MSRecordGetRequest msRecordGetRequest,
```
GET 요청에서 @RequestBody를 담지 않는 것이 바람직하다고 알고 있는데 따로 Request를 생성한 이유가 있을까요?

chech2 replied:
매개변수 개수만 생각하다보니 그 부분을 생각하지 못했던 것 같습니다!
놓쳤던 부분인데 말씀해주셔서 감사드립니다!
pathvariable로 변경했습니다!

SSUHYUNKIM commented on src/test/java/org/chzzk/howmeet/domain/regular/record/repository/TestConfig.java:
```diff
@@ -0,0 +1,21 @@
+import org.springframework.context.annotation.Bean;
+
+@TestConfiguration
+public class TestConfig {
```
TestConfig가 완전히 중복인데 하나만 둬도 괜찮지 않을까요?

chech2 replied:
파일을 어디에 둘지 고민하다가 중복으로 만들었습니다. 
global.config으로 옮기고 중복제거 했습니다. 감사합니다!!

SSUHYUNKIM commented on src/main/java/org/chzzk/howmeet/domain/common/embedded/date/impl/ScheduleDate.java:
```diff
@@ -3,11 +3,13 @@
 import java.time.LocalDate;
 
 @Embeddable
+@Getter
```
@Getter나 GetDate() 둘 중에 하나만 존재해도 되지 않나요? 혹시 따로 @Getter 를 추가한 이유가 있을까요?

chech2 replied:
네 맞습니다.  브랜치 합치기 전에 사용했는데 삭제하는게 누락된 것 같아요 수정했습니다!

chech2 committed:
[REFACTOR] private 접근 제한자 누락 수정 (#75)

chech2 committed:
[REFACTOR] nicknamesList -> nicknames 명칭 변경 (#75)

chech2 committed:
[REFACTOR] ScheduleDate의 @Getter 제거 (#75)

chech2 committed:
[REFACTOR] GSRecordNicknames, MSRecordNicknames 변경 (#75)
- convertMapToNickNameList -> distinctNicknames 메소드명 변경
- HashSet타입 set으로 변경

chech2 committed:
[REFACTOR] TestConfig 중복 제거 (#75)
- global.config로 이동 및 중복 제거

kmw2378 commented on src/main/java/org/chzzk/howmeet/domain/regular/record/dto/get/MSRecordGetResponse.java:
```diff
@@ -0,0 +1,15 @@
+import org.chzzk.howmeet.domain.common.model.SelectionDetail;
+import org.chzzk.howmeet.domain.regular.room.model.RoomName;
+
+public record MSRecordGetResponse(Long msId, RoomName roomName, Nicknames totalPersonnel,
+                                  Nicknames participatedPersonnel, List<SelectionDetail> selectTime) {
```
- Personnel 이라는 단어가 헷갈리는거 같아요! 
- 참가자 닉네임과 총 닉네임을 구분한 이유가 궁금합니다!

chech2 replied:
Personnel이라는 이름이 헷갈리기는 하는데 더 좋은 이름이 생각나지 않아 일단 해당이름으로 작성했습니다.
리팩토링 때까지 고민해보고 더 좋은 단어가 있으면 변경하도록 하겠습니다 (의견 주시면 매우 감사드리고요:) )
참가자 닉네임과 총 닉네임은 프론트 요청에 의해 구분했습니다.
해당 Ms의 모든 인원 닉네임과 현재 투표를 진행한 인원의 닉네임을 따로 받길 원하셨습니다

kmw2378 commented on src/main/java/org/chzzk/howmeet/domain/regular/record/entity/MemberScheduleRecord.java:
```diff
@@ -20,26 +19,32 @@
+    @Column(name = "room_id")
+    private Long roomId;
+
+    @Column(name = "select_time")
+    private LocalDateTime selectTime;
```
이를 VO로 가져가면 Service 레이어에서 시간 검증하기가 편할 것 같아요!

chech2 replied:
SelectionDetail, nickname 로직에 대해 리팩토링하며 한번에 같이 수정하도록 하겠습니다.

kmw2378 commented on src/main/java/org/chzzk/howmeet/domain/regular/record/service/MSRecordService.java:
```diff
@@ -0,0 +1,153 @@
+
+    @Transactional
+    public void postMSRecord(final MSRecordPostRequest msRecordPostRequest, final AuthPrincipal authPrincipal) {
+        Member member = findMemberByMemberId(authPrincipal.id());
```
`memberId`만 필요한 상황에서 SELECT 를 하여 `Member` 엔티티를 가져오는게 비효율적인 것 같아요!

chech2 replied:
member와 관련해서 한번 더 검증이 필요하지 않을까해서 추가한 부분이었는데 리뷰보고 생각해보니 필요 없는게 맞는것 같네요!!
해당 부분 수정했습니다!

chech2 committed:
[REFACTOR] nickNameMap 이름 변경 (#75)
- nicknamesById으로 이름 변경

chech2 committed:
[REFACTOR] get요청 시 responseBody 사용 수정 (#75)

chech2 committed:
[REFACTOR] 불필요한 member 엔티티 조회 부분 제거 (#75)
- member 대신 memberId 사용

chech2 committed:
[REFACTOR] 불필요한 get메소드 제거 (#75)
- getNicknames 제거

chech2 committed:
[REFACTOR] 스트림 사용 (#75)
- GSRecordNicknames, MSRecordNicknames 내부 distinctNicknames 메소드 스트림 사용한 로직으로 변경
- Nicknames 내부의 addFromNicknameSet 메소드 명 addAll로 변경

chech2 committed:
[REFACTOR]  누락된 fianl 추가 (#75)

chech2 committed:
[REFACTOR] Get 삭제에 따른 테스트 코드 수정 (#75)
- contains 추가

SSUHYUNKIM reviewed (approved):
수고하셨습니다!
